openassessment/xblock/utils/allow_resubmission.py

```diff
@@ -0,0 +1,98 @@
+"""This module contains functions that determine if a learner's assessment is eligible for resubmission."""
+
+from datetime import datetime, timedelta
+import pytz
+
+
+def allow_resubmission(config_data, workflow_data, submission_data: dict) -> bool:
+    """
+    Determines if a learner can reset their submission and try again. A learner
+    can resubmit if the following conditions are met:
+    - Whether the assignment allows learner resubmissions
+    - Whether the submission date has not been exceeded
+    - Whether the learner's response has not been graded by peers or staff
+
+    Args:
+        config_data (ORAConfigAPI): Object with all the configuration data of the ORA assignment.
+        workflow_data (WorkflowAPI): Object containing necessary data for time and workflow status checks.
+        submission_data (dict): Dictionary containing the submission data of the learner.
+
+    Returns:
+        bool: True if the learner's submission is eligible for retry, False otherwise.
+    """
+    return (
+        allow_learner_resubmissions(config_data) and not
+        submission_date_exceeded(config_data, submission_data) and not
+        has_been_graded(workflow_data)
+    )
+
+
+def allow_learner_resubmissions(config_data) -> bool:
+    """
+    Checks if the ORA assignment allows learner resubmissions.
+
+    Args:
+        config_data (ORAConfigAPI): Object with all the configuration data of the ORA assignment.
+
+    Returns:
+        bool: True if the assignment allows learner resubmissions, False otherwise
+    """
+    return config_data.allow_learner_resubmissions
+
+
+def submission_date_exceeded(config_data, submission_data: dict) -> bool:
+    """
+    Checks if the submission due date has been exceeded and if the learner is
+    within the grace period. The grace period is the time after the submission
+    date that the learner can still resubmit their response.
+
+    If the grace period is set to 0 hours and 0 minutes, means that the learner
+    can resubmit their response at any time before the submission due date.
+
+    Args:
+        config_data (ORAConfigAPI): Object with all the configuration data of the ORA assignment.
+        submission_data (dict): Dictionary containing the submission data of the learner.
+
+    Returns:
+        bool: True if the submission date has been exceeded, False otherwise.
+    """
+    ora_due_date = datetime.strptime(config_data.submission_due, "%Y-%m-%dT%H:%M:%S%z")
+    current_datetime = datetime.now(pytz.UTC)
+    ora_due_date = ora_due_date.replace(tzinfo=pytz.UTC)
+    if current_datetime >= ora_due_date:
+        return True
+
+    hours = config_data.resubmissions_grace_period_hours
+    minutes = config_data.resubmissions_grace_period_minutes
```

Comment:
What about:
```suggestion
    hours = int(config_data.resubmissions_grace_period_hours)
    minutes = int(config_data.resubmissions_grace_period_minutes)
    if not hours and not minutes:
    ...
```

Reply:
Does this parse these config data into int? https://github.com/openedx/edx-ora2/pull/2187/files#diff-99242b16512206ff1cd6b0e4e372901104bfd8307188da8b06990eb52247d284R931-R943

Reply:
It seems clearer as you suggest, I will change it. Those fields are already converted to integer

Reply:
Should we remove the casting in line 69 then?